velox/functions/prestosql/FromUtf8.cpp

```diff
@@ -0,0 +1,276 @@
+        }
+      });
+    }
+  }
+
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    return {
+        // varbinary -> varchar
+        exec::FunctionSignatureBuilder()
+            .returnType("varchar")
+            .argumentType("varbinary")
+            .build(),
+
+        // varbinary, bigint -> varchar
+        exec::FunctionSignatureBuilder()
+            .returnType("varchar")
+            .argumentType("varbinary")
+            .argumentType("bigint")
+            .build(),
+
+        // varbinary, varchar -> varchar
+        exec::FunctionSignatureBuilder()
+            .returnType("varchar")
+            .argumentType("varbinary")
+            .argumentType("varchar")
+            .build(),
+    };
+  }
+
+ private:
+  static const std::string kReplacementChar;
+
+  static std::string codePointToString(int64_t codePoint) {
+    std::string result;
+    result.resize(4);
+    stringImpl::codePointToString(result, codePoint);
+    return result;
+  }
+
+  std::string getReplacementCharacter(
+      const TypePtr& type,
+      DecodedVector& decoded,
+      vector_size_t row) const {
+    if (type->isBigint()) {
+      return codePointToString(decoded.valueAt<int64_t>(row));
+    }
+
+    auto replacement = decoded.valueAt<StringView>(row);
+    if (!replacement.empty()) {
+      auto charLength =
+          tryGetCharLength(replacement.data(), replacement.size());
+      VELOX_USER_CHECK_GT(
+          charLength, 0, "Replacement is not a valid UTF-8 character");
+      VELOX_USER_CHECK_EQ(
+          charLength,
+          replacement.size(),
+          "Replacement string must be empty or a single character")
+    }
+    return replacement;
+  }
+
+  /// Returns first row that contains invalid UTF-8 string or std::nullopt if
+  /// all rows are valid.
+  static std::optional<vector_size_t> findFirstInvalidRow(
+      const DecodedVector& decodedInput,
+      const SelectivityVector& rows) {
+    std::optional<vector_size_t> firstInvalidRow;
+    rows.testSelected([&](auto row) {
+      auto value = decodedInput.valueAt<StringView>(row);
+
+      int32_t pos = 0;
+      while (pos < value.size()) {
+        auto charLength =
+            tryGetCharLength(value.data() + pos, value.size() - pos);
+        if (charLength < 0) {
+          firstInvalidRow = row;
+          return false;
+        }
+
+        pos += charLength;
+      }
+
+      return true;
+    });
+    return firstInvalidRow;
+  }
+
+  void toVarcharNoCopy(
+      const VectorPtr& input,
+      DecodedVector& decodedInput,
+      const SelectivityVector& rows,
+      const exec::EvalCtx& context,
+      VectorPtr& result) const {
+    VectorPtr localResult;
+    if (decodedInput.isConstantMapping()) {
+      auto value = decodedInput.valueAt<StringView>(rows.begin());
+      localResult = std::make_shared<ConstantVector<StringView>>(
+          context.pool(), rows.end(), false, VARCHAR(), std::move(value));
```

Comment:
I thought we added something to prevent you from trying to move StringViews. I guess really you just want to copy it?

Reply:
@pedroerp This is required by the ConstantVector's ctor:

```
  ConstantVector(
      velox::memory::MemoryPool* pool,
      size_t length,
      bool isNull,
      TypePtr type,
      T&& val,
```